### PR TITLE
PD-205376 - Added utility method to get bv container elements from open shadow dom.

### PR DESCRIPTION
**Pull request by shrinathnayak07**
# PR description

## JIRA tickets

[PD-205376](https://bazaarvoice.atlassian.net/browse/PD-205376)

## This PR is for

- Feature 

## Requirements / issues

Added utility method to get bv container elements from open shadow dom.

## Solutions

- Added utility method to get bv container elements from open shadow dom.

## Methods

1. `shadowQuerySelectorAll` - `(node, matchSelector)` - Returns Nodes matching with match selector.
2. `findShadowRoots` - `(node)` - Returns a list of all the shadow roots found in the given root element and its descendants.
3. `filterShadowHosts` - `(node)` - Returns NodeFilter.
4. `findElementsInShadowDomByMatchSector` - `(node, matchSelector)` - Returns a NodeList of all elements within the shadow DOM of the shadowElement parameter that match the matchSelector parameter.

## Steps to verify

- using npm link to test the changes in local in bv-loader

## Check points

- [x] Functionality

## Version 
This will be released as a new minor version as we are adding a new module 

[PD-205376]: https://bazaarvoice.atlassian.net/browse/PD-205376?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/queryShadowDom/index.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  return el.querySelectorAll(matchSelector); 
+}
+
+/**
+ * Finds elements in the DOM including open shadowDOM.
+ * @param {*} element Root element to query for particular elements 
+ * @param {*} matchSelector Criteria to select the elements in the dom starting with element as root element.
+ * @returns Nodes matching with match selector.
+ */
+function shadowQuerySelectorAll (element, matchSelector) {
```
Comment (akshay-choulwar-94) [inline lib/queryShadowDom/index.js:64]: We need to cover the use case what if element or matchSelector is undefined

**Review by akshay-choulwar-94 — CHANGES_REQUESTED** [review]
Added few minor comments. Also can we try to write the unit tests for this code.

**Review by akshay-choulwar-94 — APPROVED** [review]
Looks good to me.

**Review by Shubhakar-A — APPROVED** [review]
Looks good to me

File: lib/queryShadowDom/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+| `node` | `DOM Element` | **Required**. Document node. |
+
+```javascript
+  findElementsInShadowDomByMatchSector
+  // returns a NodeList of all elements within the shadow DOM of the shadowElement parameter that match the matchSelector parameter.
+```
+| Parameter | Type     | Description                |
+| :-------- | :------- | :------------------------- |
+| `shadowElement` | `DOM Element` | **Required**. A DOM element that has a shadow root. |
+| `matchSelector` | `string` | **Required**. Criteria to select the elements in the dom starting with element as root element. |
```
Comment (soumen-saniel) [inline lib/queryShadowDom/README.md:38]: In all of the functions 1st param is a dom node.
Its better to use consistent naming `element` or `node` in all of the functions.

File: lib/queryShadowDom/index.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
+/**
```
Comment (soumen-saniel) [inline lib/queryShadowDom/index.js:1]: In this file as well have the param names consistent for dom element
Name it `element` or `node`

File: lib/queryShadowDom/index.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+* Function to deterimine whether given node is shadow host. 
+* @param {*} node Document node.
+* @returns NodeFilter 
+*/
+function filterShadowHosts (node) {
+  if (node.shadowRoot) {
```
Comment (soumen-saniel) [inline lib/queryShadowDom/index.js:7]: Better to have a null check in place `node?.shadowRoot` or `node && node.shadowRoot`

**Review by soumen-saniel — CHANGES_REQUESTED** [review]

**Review by soumen-saniel — APPROVED** [review]
Looks good

**Review by soumen-saniel — APPROVED** [review]
Looks good

**Pull request merged**